### PR TITLE
Replace initialization method by Isometry in StatePreparation

### LINKED ISSUE #12081: More efficient synthesis of StatePreparation and Initialize based on Isometry
### What should we add?

The current code for [`StatePreparation`](https://docs.quantum.ibm.com/api/qiskit/qiskit.circuit.library.StatePreparation) and [`Initialize`](https://docs.quantum.ibm.com/api/qiskit/qiskit.circuit.library.Initialize) is based on [1], but there is a more efficient code in [2] which is currently implemented in [`Isometry`](https://docs.quantum.ibm.com/api/qiskit/qiskit.circuit.library.Isometry).

Therefore, it would be better to replace the internal implementation of [`StatePreparation`](https://docs.quantum.ibm.com/api/qiskit/qiskit.circuit.library.StatePreparation) and [`Initialize`](https://docs.quantum.ibm.com/api/qiskit/qiskit.circuit.library.Initialize) by the one implemented in [`Isometry`](https://docs.quantum.ibm.com/api/qiskit/qiskit.circuit.library.Isometry).

**References**
[1] Shende, Bullock, Markov. Synthesis of Quantum Logic Circuits (2004) [https://arxiv.org/abs/quant-ph/0406176v5](https://arxiv.org/abs/quant-ph/0406176v5)
[2] Iten et al., Quantum circuits for isometries (2016). [Phys. Rev. A 93, 032318](https://journals.aps.org/pra/abstract/10.1103/PhysRevA.93.032318)

For example,
```
for n_qubits in [5, 7, 10, 12, 15]:

    state = np.random.rand(2 ** n_qubits) + np.random.rand(2 ** n_qubits) * 1j
    state = state/np.linalg.norm(state)

    # State preparation
    qc = QuantumCircuit(n_qubits)
    qc.append(StatePreparation(state), range(n_qubits))

    transpiled = transpile(qc, basis_gates=['u', 'cx'], optimization_level=0)
    cx_sp = transpiled.count_ops()['cx']

    # Isometry
    reg_iso = QuantumRegister(n_qubits)
    circuit_iso = QuantumCircuit(reg_iso)
    isom = Isometry(state, 0, 0)

    qc = QuantumCircuit(n_qubits)
    qc.append(isom, range(n_qubits))

    transpiled = transpile(qc, basis_gates=['u', 'cx'], optimization_level=0)
    cx_isom = transpiled.count_ops()['cx']

    print ("num qubits=", n_qubits, "cx count StatePreparation=", cx_sp, "cx count Isometry=", cx_isom)
```
gives:
```
num qubits= 5 cx count StatePreparation= 52 cx count Isometry= 26 
num qubits= 7 cx count StatePreparation= 240 cx count Isometry= 120 
num qubits= 10 cx count StatePreparation= 2026 cx count Isometry= 1013 
num qubits= 12 cx count StatePreparation= 8166 cx count Isometry= 4083 
num qubits= 15 cx count StatePreparation= 65504 cx count Isometry= 32752 
```

### PR DESCRIPTION
**Pull request by ShellyGarion**
### Summary

Close #12081
Replace initialization method by `Isometry` in `StatePreparation`

### CODE REVIEW

**Comment by alexanderivrii** [comment]
Thanks @ShellyGarion, this looks good. Hopefully all the tests will pass now. Could you please add the release note? And also some data on the number of gates / depth / etc. comparing the old and the new methods, i.e. highlighting improvement in this PR?

**Comment by ShellyGarion** [comment]
The following code compares the number of CX gates and circuit depth of this branch and the main branch:

```
    def test_cx_num(self):
        for n_qubits in [5, 7, 10, 12, 15]:
            state = np.random.rand(2 ** n_qubits) + np.random.rand(2 ** n_qubits) * 1j
            state = state / np.linalg.norm(state)

            # State preparation
            qc = QuantumCircuit(n_qubits)
            qc.append(StatePreparation(state), range(n_qubits))

            from qiskit import transpile
            transpiled = transpile(qc, basis_gates=['u', 'cx'], optimization_level=0)
            cx_sp = transpiled.count_ops()['cx']

            print("num qubits=", n_qubits, "cx count StatePreparation=", cx_sp, "depth StatePreparation=", transpiled.depth())
```

this branch:
```
num qubits= 5 cx count StatePreparation= 26 depth StatePreparation= 53
num qubits= 7 cx count StatePreparation= 120 depth StatePreparation= 241
num qubits= 10 cx count StatePreparation= 1013 depth StatePreparation= 2027
num qubits= 12 cx count StatePreparation= 4083 depth StatePreparation= 8167
num qubits= 15 cx count StatePreparation= 32752 depth StatePreparation= 65505
```
main branch:
```
num qubits= 5 cx count StatePreparation= 52 depth StatePreparation= 110
num qubits= 7 cx count StatePreparation= 240 depth StatePreparation= 488
num qubits= 10 cx count StatePreparation= 2026 depth StatePreparation= 4063
num qubits= 12 cx count StatePreparation= 8166 depth StatePreparation= 16345
num qubits= 15 cx count StatePreparation= 65504 depth StatePreparation= 131024
```

**Comment by alexanderivrii** [comment]
Very nice, we see 2x reduction in the number of gates. Not directly related to this PR, but do you think it is worthwhile to track state preparation benchmarks via ASV (in the case that they are not there already)?

**Comment by ShellyGarion** [comment]
> Very nice, we see 2x reduction in the number of gates. Not directly related to this PR, but do you think it is worthwhile to track state preparation benchmarks via ASV (in the case that they are not there already)?

I think it would be nice to add it. I think that `Isometry` is being tracked in: https://github.com/Qiskit/qiskit/blob/main/test/benchmarks/isometry.py, but it would be nice to add a specific case for `StatePreparation`.

File: qiskit/circuit/library/generalized_gates/isometry.py [hunk 1]
```diff
@@ -139,8 +139,8 @@ def _gates_to_uncompute(self):
         Call to create a circuit with gates that take the desired isometry to the first 2^m columns
          of the 2^n*2^n identity matrix (see https://arxiv.org/abs/1501.06911)
         """
-        q = QuantumRegister(self.num_qubits)
-        circuit = QuantumCircuit(q)
+        q = QuantumRegister(self.num_qubits, "q")
+        circuit = QuantumCircuit(q, name="isometry")
```
Comment (alexanderivrii) [inline qiskit/circuit/library/generalized_gates/isometry.py]: This is not the full circuit, correct? Should we possibly call it something like "isometry_to_uncompute"?

**Review by alexanderivrii — COMMENTED** [review]
LGTM! Just one small comment: I am personally in favor of removing internal code that is not likely to be ever used in the future. @ShellyGarion, you don't think that any of the removed methods in statepreparation.py will ever be useful, correct?

**Comment by alexanderivrii** [comment]
A quick run of ASV on the new statepreparation benchmark:
```
t have improved:

| Change   |   Before [b5c5179a] <main> |   After [4dab85c4] <pr-shelly> |   Ratio | Benchmark (Parameter)                                                                                   |
|----------|----------------------------|--------------------------------|---------|---------------------------------------------------------------------------------------------------------|
| -        |                         28 |                             17 |    0.61 | statepreparation.StatePreparationTranspileBench.track_cnot_counts_after_mapping_to_ibmq_16_melbourne(4) |
| -        |                        186 |                            111 |    0.6  | statepreparation.StatePreparationTranspileBench.track_cnot_counts_after_mapping_to_ibmq_16_melbourne(6) |
| -        |                         79 |                             47 |    0.59 | statepreparation.StatePreparationTranspileBench.track_cnot_counts_after_mapping_to_ibmq_16_melbourne(5) |
| -        |                        411 |                            231 |    0.56 | statepreparation.StatePreparationTranspileBench.track_cnot_counts_after_mapping_to_ibmq_16_melbourne(7) |
| -        |                        872 |                            460 |    0.53 | statepreparation.StatePreparationTranspileBench.track_cnot_counts_after_mapping_to_ibmq_16_melbourne(8) |

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.


```

**Review by alexanderivrii — DISMISSED** [review]

File: qiskit/circuit/library/generalized_gates/isometry.py (hunk 1 above)
Comment (ShellyGarion) [reply qiskit/circuit/library/generalized_gates/isometry.py]: Thanks, I've updated the name.
I also removed all the old synthesis code from the statepreparation class.

**Review by alexanderivrii — APPROVED** [review]

**Pull request merged**